File: src/frontend/services/apiClients/chapters.ts
// Copyright (C) 2026 StableLlama
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.

/**
 * Defines the chapters unit so this responsibility stays isolated, testable, and easy to evolve.
 */

import { Conflict } from '../../types';
import { ChapterDetailResponse, ChapterListResponse } from '../apiTypes';
import { ContentWriteOptions, RevisionedContent } from '../contentRevision';
import { fetchJson, putJson, postJson, deleteJson, projectEndpoint } from './shared';

export type RevisionedChapterDetailResponse = ChapterDetailResponse &
  RevisionedContent & {
    book_id?: string | null;
  };

export interface ChapterContentWriteResponse extends RevisionedContent {
  ok: boolean;
  id: number;
}

export interface ChaptersApi {
  list: () => Promise<ChapterListResponse>;
  get: (id: number) => Promise<RevisionedChapterDetailResponse>;
  create: (
    title: string,
    content?: string,
    book_id?: string
  ) => Promise<{
    ok: boolean;
    id: number;
    title: string;
    book_id?: string | undefined;
  }>;
  updateContent: (
    id: number,
    content: string,
    options?: ContentWriteOptions
  ) => Promise<ChapterContentWriteResponse>;
  updateTitle: (id: number, title: string) => Promise<{ ok: boolean }>;
  updateSummary: (id: number, summary: string) => Promise<{ ok: boolean }>;
  updateMetadata: (
    id: number,
    data: {
      summary?: string;
      notes?: string;
      private_notes?: string;
      conflicts?: Conflict[];
    }
  ) => Promise<{ ok: boolean; id?: number | undefined; message?: string | undefined }>;
  delete: (id: number) => Promise<{ ok: boolean }>;
  reorder: (chapterIds: number[], bookId?: string) => Promise<{ ok: boolean }>;
}

export const createChaptersApi = (projectName: string): ChaptersApi => ({
  list: async () =>
    fetchJson<ChapterListResponse>(
      projectEndpoint(projectName, '/chapters'),
      undefined,
      'Failed to list chapters'
    ),

  get: async (id: number): Promise<RevisionedChapterDetailResponse> => {
    return fetchJson<RevisionedChapterDetailResponse>(
      projectEndpoint(projectName, `/chapters/${id}`),
      undefined,
      'Failed to get chapter'
    );
  },

  create: async (
    title: string,
    content: string = '',
    book_id?: string
  ): Promise<{
    ok: boolean;
    id: number;
    title: string;
    book_id?: string | undefined;
  }> => {
    return postJson<{ ok: boolean; id: number; title: string; book_id?: string }>(
      projectEndpoint(projectName, '/chapters'),
      { title, content, book_id },
      'Failed to create chapter'
    );
  },

  updateContent: async (
    id: number,
    content: string,
    options?: ContentWriteOptions
  ): Promise<ChapterContentWriteResponse> => {
    return putJson<ChapterContentWriteResponse>(
      projectEndpoint(projectName, `/chapters/${id}/content`),
      { content, ...options },
      'Failed to update chapter content'
    );
  },

  updateTitle: async (id: number, title: string): Promise<{ ok: boolean }> => {
    return putJson<{ ok: boolean }>(
      projectEndpoint(projectName, `/chapters/${id}/title`),
      { title },
      'Failed to update chapter title'
    );
  },

  updateSummary: async (id: number, summary: string): Promise<{ ok: boolean }> => {
    return putJson<{ ok: boolean }>(
      projectEndpoint(projectName, `/chapters/${id}/summary`),
      { summary },
      'Failed to update chapter summary'
    );
  },

  updateMetadata: async (
    id: number,
    data: {
      summary?: string;
      notes?: string;
      private_notes?: string;
      conflicts?: Conflict[];
    }
  ): Promise<{
    ok: boolean;
    id?: number | undefined;
    message?: string | undefined;
  }> => {
    return putJson<{ ok: boolean; id?: number; message?: string }>(
      projectEndpoint(projectName, `/chapters/${id}/metadata`),
      data,
      'Failed to update chapter metadata'
    );
  },

  delete: async (id: number): Promise<{ ok: boolean }> => {
    return deleteJson<{ ok: boolean }>(
      projectEndpoint(projectName, `/chapters/${id}`),
      'Failed to delete chapter'
    );
  },

  reorder: async (chapterIds: number[], bookId?: string): Promise<{ ok: boolean }> => {
    return fetchJson<{ ok: boolean }>(
      projectEndpoint(projectName, '/chapters/reorder'),
      {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(
          bookId
            ? { book_id: bookId, chapter_ids: chapterIds }
            : { chapter_ids: chapterIds }
        ),
      },
      'Failed to reorder chapters'
    );
  },
});

export const chaptersApi = createChaptersApi('');

File: src/augmentedquill/models/chapters.py
# Copyright (C) 2026 StableLlama
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.

"""Defines the chapters unit so this responsibility stays isolated, testable, and easy to evolve.

Pydantic models for chapter-related API responses.
"""

from __future__ import annotations

from typing import Any

from pydantic import BaseModel, ConfigDict


class ChapterSummary(BaseModel):
    """Lightweight chapter descriptor used in list responses."""

    id: int
    title: str
    filename: str
    summary: str
    notes: str
    private_notes: str
    conflicts: list[Any]
    book_id: str | None = None
    document_key: str | None = None
    source_path: str | None = None
    manuscript_status: str | None = None


class ChaptersListResponse(BaseModel):
    """Response body for ``GET /api/v1/chapters``."""

    chapters: list[ChapterSummary]


class ChapterDetailResponse(BaseModel):
    """Response body for ``GET /api/v1/chapters/{chap_id}``."""

    id: int
    title: str
    filename: str
    content: str
    summary: str
    notes: str
    private_notes: str
    conflicts: list[Any]
    book_id: str | None = None
    revision: str
    document_key: str
    source_path: str | None = None
    manuscript_status: str | None = None


class ChapterMetadataUpdate(BaseModel):
    """Request body for updating chapter metadata."""

    title: str | None = None
    summary: str | None = None
    notes: str | None = None
    private_notes: str | None = None
    conflicts: list[Any] | None = None


class ChapterTitleUpdate(BaseModel):
    """Request body for updating chapter title."""

    title: str


class ChapterCreate(BaseModel):
    """Request body for creating a new chapter."""

    title: str
    content: str | None = ""
    book_id: str | None = None


class ChapterContentUpdate(BaseModel):
    """Request body for updating chapter content."""

    content: str
    expected_revision: str | None = None
    expected_filename: str | None = None
    expected_document_key: str | None = None


class ChapterSummaryUpdate(BaseModel):
    """Request body for updating chapter summary."""

    summary: str


class ChaptersReorderRequest(BaseModel):
    """Request body for reordering chapters."""

    model_config = ConfigDict(extra="forbid")

    chapter_ids: list[int]
    book_id: str | None = None


class BooksReorderRequest(BaseModel):
    """Request body for reordering books."""

    model_config = ConfigDict(extra="forbid")

    book_ids: list[str]
